### ISSUE TITLE
Invalid public key in parsed X509 certificates

### ISSUE THREAD

**Issue by evolutionerp** [Ephenodrom/Dart-Basic-Utils#76]
Hello,

After parsing the following certificate with the [x509CertificateFromPem](https://pub.dev/documentation/basic_utils/latest/basic_utils/X509Utils/x509CertificateFromPem.html) function:

```
-----BEGIN CERTIFICATE-----
MIIEEzCCAvugAwIBAgIJAJNxDw+7d5AvMA0GCSqGSIb3DQEBCwUAMIGUMUowSAYD
VQQDE0FTZXJ2aXppbyBiYWNrYm9uZSBkaSBFdm9sdXRpb24gKDBkMDc0NGMxYWMw
NzRhNWM4NDMwZGZiZjc4MTBhYTNmKTFGMEQGCSqGSIb3DQEJARY3MGQwNzQ0YzFh
YzA3NGE1Yzg0MzBkZmJmNzgxMGFhM2ZAY3VzdG9tZXJzLmV2b2x1dGlvbi5pdDAe
Fw0yMjA5MTExMjI1MTZaFw0yMjEwMTAxMjI1MTZaMIGUMUowSAYDVQQDE0FTZXJ2
aXppbyBiYWNrYm9uZSBkaSBFdm9sdXRpb24gKDBkMDc0NGMxYWMwNzRhNWM4NDMw
ZGZiZjc4MTBhYTNmKTFGMEQGCSqGSIb3DQEJARY3MGQwNzQ0YzFhYzA3NGE1Yzg0
MzBkZmJmNzgxMGFhM2ZAY3VzdG9tZXJzLmV2b2x1dGlvbi5pdDCCASIwDQYJKoZI
hvcNAQEBBQADggEPADCCAQoCggEBAMHVQOYMkUPDv+LGFhOuc/LrZ1Bui06TfqJP
1X9ADLT9O9Sp2iK6KFelxV640wJgezj1pDwH7Bf6Mkcpfjpt/M42sdqX/R0blhCy
2mpTzn9Rp10eQHNMenj/V7768jsYb6qfPLQ/pR8g9IGNGUIuUYpa9JCcACwzsHQB
SMEjhwfyjO8XqjS69y6ZO4iYvPbrwgtL9jM0rZEwI/y9fud75D6ZkpC0Es+athBI
KrBgaGzpibJt/uo21h+n2zDP+Q14kL64c2PiaGWildYUOlpL28bIbvXk1mgkw1NR
0vXkHgNv4ymgkXhrn1uJBbTK11kL3sCcJExzmgSMhbDzSoDw3O0CAwEAAaNmMGQw
PQYDVR0RBDYwNIcEfwAAAYcQAAAAAAAAAAAAAAAAAAAAAYIJbG9jYWxob3N0ggRE
RVY1gglkZXY1LmRldjUwDgYDVR0PAQH/BAQDAgSwMBMGA1UdJQQMMAoGCCsGAQUF
BwMBMA0GCSqGSIb3DQEBCwUAA4IBAQCKFUYRUo9iavEDrBtrkPM2MSnU9nhWu52E
lgnL4xijWYPbdDBftjNF3gz1/zEh/h+YyUDO/zO9vnKzoZS0xrJPE2olnwz5jE4e
1nuCfsM8uswKcKEiERmmJlgSyL409rZcfXmKopbVSVqCUTEc/D723Bk4knTXR6Mg
HhjG2/VCxCr6297K+0BAB5msyMPO5J9OpNp0LQt3MPFBBYyUuAlcqSI+O+MAjxA7
QR7rWnbyXLNLh2sioNjigi2EgOFy4JeDQBFq3+t8s4sGK3GssIVvdZ/vFrnVHjqF
oFIghMrMIQVBCAorC+Q7T5fqtb8bUfRfqOaU5okvLT6G5rLv+QoW
-----END CERTIFICATE-----
```

In version **4.5.2** the expected public key is accessible via `.publicKeyData.bytes`:

```
3082010A0282010100C1D540E60C9143C3BFE2C61613AE73F2EB67506E8B4E937EA24FD57F400CB4FD3BD4A9DA22BA2857A5C55EB8D302607B38F5A43C07EC17FA3247297E3A6DFCCE36B1DA97FD1D1B9610B2DA6A53CE7F51A75D1E40734C7A78FF57BEFAF23B186FAA9F3CB43FA51F20F4818D19422E518A5AF4909C002C33B0740148C1238707F28CEF17AA34BAF72E993B8898BCF6EBC20B4BF63334AD913023FCBD7EE77BE43E999290B412CF9AB610482AB060686CE989B26DFEEA36D61FA7DB30CFF90D7890BEB87363E26865A295D6143A5A4BDBC6C86EF5E4D66824C35351D2F5E41E036FE329A091786B9F5B8905B4CAD7590BDEC09C244C739A048C85B0F34A80F0DCED0203010001
```
(270 bytes, sha-256: `97c07fc52a6409f376c42bd3310aa1b7b84356249ccff6da9afcb68b8f40f3b3`)

---
However, in version **5.2.2** (and **5.2.1**) an invalid blob is returned instead (both via `.publicKeyData.bytes` and `.tbsCertificate.subjectPublicKeyInfo.bytes`):

```
30820122300D06092A864886F70D01010105000382010F003082010A0282010100C1D540E60C9143C3BFE2C61613AE73F2EB67506E8B4E937EA24FD57F400CB4FD3BD4A9DA22BA2857A5C55EB8D302607B38F5A43C07EC17FA3247297E3A6DFCCE36B1DA97FD1D1B9610B2DA6A53CE7F51A75D1E40734C7A78FF57BEFAF23B186FAA9F3CB43FA51F20F4818D19422E518A5AF4909C002C33B0740148C1238707F28CEF17AA34BAF72E993B8898BCF6EBC20B4BF63334AD913023FCBD7EE77BE43E999290B412CF9AB610482AB060686CE989B26DFEEA36D61FA7DB30CFF90D7890BEB87363E26865A295D6143A5A4BDBC6C86EF5E4D66824C35351D2F5E41E036FE329A091786B9F5B8905B4CAD7590BDEC09C244C739A048C85B0F34A80F0DCED0203010001A3663064303D0603551D110436303487047F00000187100000000000000000000000000000000182096C6F63616C686F73748204444556358209646576352E64657635300E0603551D0F0101FF0404030204B030130603551D25040C300A06082B06010505070301
```
(398 bytes, sha-256: `17944b6375f0b0f268f8ca2a524d63063d08ee80b5d8f173f50597a9aa31f0ad`)

---
This unfortunately breaks public-key pinning in my application, forcing me to keep a dependency on the old version of the package. Am I doing something wrong?

**Comment by Ephenodrom** [comment]
Hello @evolutionerp 

I will take a look at it later! You can now stay at the older version, that should be no problem, due to the changes since that are mostly new features or you use the current version and access the deprecated field.

I keep you updated!

Regards

**Comment by evolutionerp** [comment]
> use the current version and access the deprecated field
Also the deprecated field gets populated with the invalid key, unfortunately.

**Comment by Ephenodrom** [comment]
@evolutionerp 
After digging a bit in the older code from 4.4.2 I found out that the filling of the bytes value has changed with 5.0.0. The old way was not correct and with the old way it was not possible to create a public key out of the bytes value. I improved the code and made some changes to reuse some code that is similar within CSR and X509 parsing. The public key parsing of the a CSR was already correct so I reused the code now for X509.

In the old way only the SEQUENCE with the two INTEGERS was set in the byte list.
![grafik](https://user-images.githubusercontent.com/30625794/190385697-18098bd0-8b11-4985-bf1c-30711b643424.png)

Now the bytes contain the whole public key SEQUENCE.

Now there are several ways. 

1) Parse the key from the bytes as it is done in the CryptoUtils class with rsaPrivateKeyFromDERBytes().

Example that would return the initial bytes from the new bytes.
```dart
  static Uint8List bytes rsaPrivateKeyFromDERBytes(Uint8List bytes) {
    var asn1Parser = ASN1Parser(bytes);
    var topLevelSeq = asn1Parser.nextObject() as ASN1Sequence;
    //ASN1Object version = topLevelSeq.elements[0];
    //ASN1Object algorithm = topLevelSeq.elements[1];
    var privateKey = topLevelSeq.elements![2];

    asn1Parser = ASN1Parser(privateKey.valueBytes);
    var pkSeq = asn1Parser.nextObject() as ASN1Sequence;
    return pkSeq.encodedBytes!;
  }
```

2) Use the new bytes to compare the public key
3) Use the sha256Thumbprint field to compare the public key.

I hope that this helps!

Regards
